test/token/error_message_test.dart

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// @dart=2.10
+
+import 'package:shelf/shelf.dart' as shelf;
+import 'package:test/test.dart';
+
+import '../descriptor.dart' as d;
+import '../test_pub.dart';
+
+void respondWithWwwAuthenticate(String headerValue) {
+  globalPackageServer.expect('GET', '/api/packages/versions/new', (request) {
+    return shelf.Response(403, headers: {'www-authenticate': headerValue});
+  });
+}
+
+Future<void> expectPubErrorMessage(dynamic matcher) {
+  return runPub(
+    args: ['lish'],
+    environment: {
+      'PUB_HOSTED_URL': globalPackageServer.url,
+      'PUB_TEST_AUTH_METHOD': 'token',
+    },
+    exitCode: 65,
+    input: ['y'],
+    error: matcher,
+  );
+}
+
+void main() {
+  setUp(() async {
+    await d.validPackage.create();
+    await servePackages();
+    await d.tokensFile({
+      'version': 1,
+      'hosted': [
+        {'url': globalPackageServer.url, 'token': 'access token'},
+      ]
+    }).create();
+  });
+
+  test('prints www-authenticate message', () async {
+    respondWithWwwAuthenticate('bearer realm="pub", message="custom message"');
+    await expectPubErrorMessage(contains('custom message'));
+  });
+
+  test('sanitizes and prints dirty www-authenticate message', () {
+    // Unable to test this case because shelf does not allow characters [1]
+    // that pub cli supposed to sanitize.
+    //
+    // [1] https://github.com/dart-lang/sdk/blob/main/sdk/lib/_http/http_headers.dart#L653-L662
+  });
+
+  test('trims and prints dirty www-authenticate message', () async {
+    var message = List.generate(2048, (_) => 'a').join();
+
+    respondWithWwwAuthenticate('bearer realm="pub", message="$message"');
+    await expectPubErrorMessage(contains(message.substring(0, 1024)));
```

Comment:
This is a useless test... you need to test that it does not contain `message`... if the 2048 chars long message is printed, then it'll still contain a 1024 substring of the message.

Reply:
ah ok